### DISCUSSION TITLE
$onAction not always firing?

### DISCUSSION THREAD

**Discussion by mixmix** [vuejs/pinia#1400] [Help and Questions] [answered]
### Reproduction

https://stackblitz.com/edit/github-cktvx5-g4kkro?file=src/stores/index.js

### Steps to reproduce the bug

Setup:
- quasar 2.7.3 with vite
- vue 3.2.37
- pinia 2.0.14

I'm writing an undo-redo plugin, and it is _semi_ working. My approach is subscribing to actions, then recording those. The problem is that only some actions seem to "heard"... and worse still it's intermittent :cry: 

Plugin code so far, any advice most appreciated: 

```js
export default store((/* { ssrContext } */) => {
  const pinia = createPinia()

  // add Pinia plugins
  pinia.use(history)

  return pinia
})

const TIMETRAVEL = 'timetravel'
function history ({ store, options }) {
  if (!options.history) return

  const past = reactive([])
  const future = reactive([])

  store.$onAction(action => {
    // do not record timetravel actions
    if (isTimetravel(action)) return

    past.push({ name: action.name, payload: action.args })
    while (future.length) future.pop()
  })

  const replay = () => {
    store.$reset()
    past.forEach(action => store[action.name](...action.payload, TIMETRAVEL))
  }

  return {
    past,
    canUndo: computed(() => past.length > 0),
    canRedo: computed(() => future.length > 0),
    undo () {
      const action = past.pop()
      future.push(action)

      replay()
    },
    redo () {
      const action = future.pop()
      past.push(action)

      replay()
    }
  }
}

function isTimetravel (action) {
  return action.args[action.args.length - 1] === TIMETRAVEL
}

``` 

### Expected behavior

the $onAction callback should fire ever time, on all actions for stores which "turn on" history

### Actual behavior

When I try to use this history plugin in a store, I see one action triggering onAction almost always, and one almost never
```js
import { defineStore } from 'pinia'

export const useNetworkDataStore = defineStore('network/data', {
  state: () => ({
    seedNodes: {
      // nodeId: data..
    },
    directedLinks: {
      // A: { B: true }
    }
  }),
  getters: {
    // ...
  },
  actions: {
    addNode (node) {
      // this one is always heard
    },

    addDirectedLinks (links) {
      // this one is almost never heard
    }
  },

  history: true
  // WARNING - records and replays *actions*
  // DO NOT put side-effects in actions, otherwise these will all be replayed!
})
```

**Comment by mixmix** [comment]
@posva is this not a possible bug?
I did make a repo trying to make a simple plugin to demo the problem I'm seeing with $onAction

If I misread the docs and used the plugin system wrong that would be great (ie no bug).
Can you clarify if that's the case?

**Reply by guyaumetremblay** [reply]
@mixmix I know this is an old bug / discussion but did you find the cause? I got the same thing but different... I have 2 plugins, each one subscribing to $onAction and each one doing a simple console.log of the name... but sometime, only one of both is loging !

**Reply by posva** [reply, marked as answer]
I completely missed this 👀 

The problem in the repro is that no action is being called. [It works when calling the actions](https://play.pinia.vuejs.org/#eNqNVl1vIjcU/SteVIkhBU+iVFuJQpQ0i9pU3SRiSfuw5GEyY8DZGdsae4Ao4r/32tczY0LK7hNj+36ce+7xNa+dK6XoumKdYWek05IrQzQzlSJ5IpbjecfoeediLnihZGnIK6k0u5aVMKwkO7IoZUG6NE5xhxrdbU0TpfKX+7xactFarrg2snwZKLePDnMRmEY9u5FKoQ3xYck4yGrPRzEiBVywMKxQeWIYrAgZqYvZijWeXJMEoLzWG1SQ3W4UK7R9qoyRglymOU+/Qa21ERdpyQomTHTWg+pv6uUoRo/j3hnb8/5ULwNv55/xtQt0JFQlMgkFdy4ebj/dhWWoRBuaM7E0K1dQgOtIuJL5cNPJfrhFZaqS/U/AUYxIR3HAdKffwT4PikTRZy0FCOjVms/9AehmSNyO3VNc8MTuzDux+6ZMF4OnUm405H8G435teglqjDO2NlLmepAojm4rY5QexnGaCbDPWM7X0E5mYqGK+MDn8iP9SM/inD/FkCjmImPb/TTgMchY8SPRa9PLU3r2C/3VRV2f+6CFjWqD7uZiB7QYDeJd8OUbUlJZKJ6z8k4ZDuLeIyfJc7n5y+2ZsmINxHTF0m/v7D/rLaK+LxnQt2ZBWSYpl8zg8eTLLdvCd3NYyKzKwfrI4ZRpmVcWI5r9DhoE2IGdQ3vjeszFcqYnW8OErouyQB0bzt7RfH2k9BbuOT0PWGwnClDYDJ+MLbhgX2CCsD5J0pQp8+fn6YPKQJTNOHLy6v5mbxnbOkecJsHkGoeRoq5P1u0jLm0g3JBEPTK+IJGHKobk1JW068GP/UhSV05TTDs1XLxevU+IgakHk2dcf/yMFw8g2lPLFH61k+N7EQYHEdo4fsQGmolj8u/V9Pbm9g9wLFkqywzmosjgG+7ziyYnvpYTbw3T4fZuRlQFjwHP2IAtFiw1Gkqsq+4TaVas3HDNABOokGx4nhPQBnliPi7LPkAze64TfEEibCMtmEnoStblQbo3BxQ7G71pcNQ2sB+69GwGkA2o5uB1CcUD3EKUe6uOPgHRXUEla1wfaqfxAidn1wfGCyCEZY0xSPsdmQVvGfQrik8giNbltQTgWwPOJ7ETliseXVxWMN4HhS+hIyjJMoJAsTJt93GMAimRL9vyQAAwzHKBp5YYhxDzzG4+T2bTq38mf0OyruHAXpmsWW6rWFTCdbaWTwSo8Z5JvLYgfMRsW/nBb9I6tU/rkjVlwSMFiWoGo6+PiBBP8c05PMcbCJnpT1JYOqSIQAr2t+UNg8gc3iy5jLp4DPcXP6hICjcg3AImjMbEjstMEiGNvwakJaFWtr/MVq961pzWCN4UCl2wL7Gq9AoIs2mH+xhU8pLLJGt2LRhg0sPZrGAukmjv+e15ZqiSKkJDf4dq5vB2WXGFhHjO7ItgvJ9Ht5DlJElXeyw6668B1MeIUrh4bu1B9wPBuEsGSDwQLzKf2mbxzwNi94s0EQ/wD2bYXJ0IEYf/Xi7IqR2o3nwK/1AOzPf/nQQO+PeonZL+AqKOx741DYstPOyXZ6PppK3KEtuY41i120fTHLRrXxY/mMYyC0MsuIjvyQ9BePYDSdWdtN81UQNy9kjG43HQRT8od/8BrlIGVQ==).

A limitation you might have encountered is that `$onAction()` cannot be aware of actions invoked by other actions **in setup stores**. This is a technical limitation that needs some investigation but should be resolved in a new major. @guyaumetremblay I recommend you to try and boil it down in the playground to see if it's a bug

FYI: issues are moved to discussions when they don't follow the reporting guidelines (instructions in "new issue" page). In your case, I can't remember anymore but it was probably that this is just a usage question with an editable example not a bug report.

**Reply by guyaumetremblay** [reply]
Hi @posva and thanks for your reply. 
I think that my situation is different. I'll try to create a small repro but here is a more detailed but simplified explanation: 
- Take a component with a button calling a pinia action. 
- Add 2 plugins, each using $onAction() to log in the console the name of the action. 

In my case, only one of both plugins is printing the log! I don't know why a limitation would affect one plugin but not the second?

**Reply by guyaumetremblay** [reply]
Nevermind, I just found my issue... Don't know if it's something missing in the documentation. 
I just understand that a plugin is installed **for each store** ! So, if you have 2 stores, the install function will be called twice; not only once with the root store but twice with each store. 
My second plugin was having a guard, to install it only once so it was not listening the actions of the second store.

**Reply by posva** [reply]
Ah yeah, you can use guards but they should be for each plugin 😄 
Each plugin is independent so I would say this makes sense without any extra documentation

**Comment by posva** [comment]
An update on this: I ~~plan on adding~~ added an `action` helper to setup stores that allows users to mark functions as actions:

```ts
defineStore('id', ({ action }) => {
  const n = ref(0)
  const increment = action((amount = 1) => n.value += amount)
  const decrement = action((amount) => increment(-amount))
  return { n, increment }
})
```

With the setup above, calling `store.decrement(1)` will trigger `$onAction()` callbacks twice: one for `decrement` and one for `increment`.

This approach is opt-in, everything will work the same without it. It feels a bit clunky but it's not invasive and really simple without the need of a build tool. Feedback welcome of course

**Reply by ddenev** [reply]
@posva , I just found this post since I had the same problem - calling an action in the same store and expecting this to fire $onAction in another store (subscribee) but it did not work.

So my questions are:
- since I cannot find this (the `action` helper) in the documentation, is this a public API now and will it be documented?
- is this the only use case for `action` - to be able to call an action in the same store and trigger subscriptions? If so, do we need it, since we always can do `useSameStore().myAction()` and this seems to work.
- can you add this 'quirk' to the docs as well - it took me couple of hours to actually find the reason why my subscriptions were not called 🙂

**Comment by stavrosdro** [comment]
Regarding the current situation when I want to call an action from the same store and want to trigger `$onAction` for both of them, can I follow the approach that I did [here](https://stackblitz.com/edit/vitejs-vite-mueawv?file=src/stores/counterV2.ts) ? Is it safe?
